lib/charms/hpc_libs/v0/slurm_ops.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+"""Library to manage the Slurm snap.
+
+This library contains the `SlurmManager` class, which offers interfaces to use and manage
+the Slurm snap inside charms.
+
+### General usage
+
+For starters, the `SlurmManager` constructor receives a `Service` enum as a parameter, which
+helps the manager determine things like the correct service to enable, or the correct settings
+key to mutate.
+
+```
+from charms.hpc_libs.v0.slurm_ops import (
+    Service,
+    SlurmManager,
+)
+
+class ApplicationCharm(CharmBase):
+    # Application charm that needs to use the Slurm snap.
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+        # Charm events defined in the NFSRequires class.
+        self._slurm_manager = SlurmManager(Service.SLURMCTLD)
+        self.framework.observe(
+            self.on.install,
+            self._on_install,
+        )
+
+    def _on_install(self, _) -> None:
+        self._slurm_manager.install()
+        self.unit.set_workload_version(self._slurm_manager.version())
+        self._slurm_manager.set_config("cluster-name", "cluster")
+```
+"""
+
+import logging
+import functools
+import subprocess
+import yaml
+import base64
+import enum
+import tempfile
+import os
+
+
+_logger = logging.getLogger(__name__)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "541fd767f90b40539cf7cd6e7db8fabf"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+def _call(cmd: str, *args: [str]) -> bytes:
+    """Call a command with logging.
+
+    Raises:
+        subprocess.CalledProcessError: Raised if the command fails.
+    """
+    cmd = [cmd, *args]
+    _logger.debug(f"Executing command {cmd}")
+    try:
+        return subprocess.check_output(cmd, stderr=subprocess.PIPE, text=False)
```

Comment:
I suspected that was your reasoning; you would use `read_bytes(...)` if reading the munge key file directly in Python. 

Tbh I was thinking we make `getmungekey` and `setmungekey` take a base64 value directly and handle the decoding/encoding. That way all we have to worry about within the charms is just exchanging the string value back and forth via the relation data buckets. I personally find base64 strings much easier to work with compared to raw bytes objects :sweat_smile: 

However, we can go forward with this implementation you have here since we haven't updated the API for managing the munge key in the Slurm snap yet.